File: frontend/app/routes/settings/usenet/provider-connection-budget.ts
export type ProviderConnectionBudgetPreview = Readonly<{
  providerLimit: number;
  transferLimit: number;
  baseMetadataCapacity: number;
  metadataBurstAllowance: number;
  maxMetadataCapacity: number;
}>;

export function calculateProviderConnectionBudget(
  providerLimit: number,
  transferLimit: number,
): ProviderConnectionBudgetPreview | null {
  if (
    !Number.isInteger(providerLimit) ||
    providerLimit <= 0 ||
    !Number.isInteger(transferLimit) ||
    transferLimit <= 0 ||
    transferLimit > providerLimit
  ) {
    return null;
  }

  const baseMetadataCapacity = providerLimit - transferLimit;
  const calculatedMetadataMax = baseMetadataCapacity + Math.floor(transferLimit / 2);
  const maxMetadataCapacity = Math.min(providerLimit, Math.max(1, calculatedMetadataMax));
  const metadataBurstAllowance = maxMetadataCapacity - baseMetadataCapacity;

  return {
    providerLimit,
    transferLimit,
    baseMetadataCapacity,
    metadataBurstAllowance,
    maxMetadataCapacity,
  };
}

export function formatMetadataCapacity(
  budget: Pick<ProviderConnectionBudgetPreview, "baseMetadataCapacity" | "maxMetadataCapacity">,
): string {
  return `${budget.baseMetadataCapacity}\u2013${budget.maxMetadataCapacity} connections`;
}

File: backend/Clients/Usenet/Connections/ProviderConnectionBudget.cs
namespace NzbWebDAV.Clients.Usenet.Connections;

/// <summary>
/// Effective per-provider transfer and metadata limits for split connection scheduling.
/// A null configured transfer limit is legacy mode and therefore does not produce a budget.
/// </summary>
internal readonly record struct ProviderConnectionBudget(
    int EffectiveProviderLimit,
    int EffectiveTransferLimit,
    int BaseMetadataCapacity,
    int MetadataBurstAllowance,
    int MaxMetadataCapacity)
{
    public static ProviderConnectionBudget Calculate(
        int effectiveProviderLimit,
        int configuredTransferLimit)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(configuredTransferLimit);

        var providerLimit = Math.Max(1, effectiveProviderLimit);
        var transferLimit = Math.Min(configuredTransferLimit, providerLimit);
        var baseMetadataCapacity = providerLimit - transferLimit;

        // A one-connection provider still needs an opportunistic metadata lease for
        // STAT/HEAD/DATE and recovery probes while its transfer slot is idle.
        var calculatedMetadataMax = baseMetadataCapacity + transferLimit / 2;
        var maxMetadataCapacity = Math.Min(providerLimit, Math.Max(1, calculatedMetadataMax));
        var metadataBurstAllowance = maxMetadataCapacity - baseMetadataCapacity;

        return new ProviderConnectionBudget(
            providerLimit,
            transferLimit,
            baseMetadataCapacity,
            metadataBurstAllowance,
            maxMetadataCapacity);
    }
}
